File: ios_app/iphone/RepeatingFormPeriodFs.swift
import SwiftUI
import shared

struct RepeatingFormPeriodFs: View {

    @State private var vm: RepeatingFormPeriodVm
    @Binding private var isPresented: Bool
    private let onPick: (RepeatingDbPeriod?) -> Void

    // Needed for SwiftUI picker, up via onChange(). Set in onAppear().
    @State private var pickerNDaysTemp = 2
    @State private var isAddDayOfYearSheetPresented = false

    init(
        isPresented: Binding<Bool>,
        defaultPeriod: RepeatingDbPeriod?,
        onPick: @escaping (RepeatingDbPeriod?) -> Void
    ) {
        _vm = State(initialValue: RepeatingFormPeriodVm(defaultPeriod: defaultPeriod))
        _isPresented = isPresented
        self.onPick = onPick
    }

    var body: some View {

        VMView(vm: vm, stack: .VStack()) { state in

            Fs__HeaderAction(
                title: state.title,
                actionText: state.doneText,
                scrollToHeader: 0,
                onCancel: {
                    isPresented = false
                },
                onDone: {
                    vm.buildSelectedPeriod { period in
                        onPick(period)
                        isPresented = false
                    }
                }
            )

            MyListView__PaddingFirst()

            ForEach(0..<state.periods.count, id: \.self) { periodIndex in

                let isFirst = periodIndex == 0

                MyListView__ItemView(
                    isFirst: isFirst,
                    isLast: state.periods.count == periodIndex + 1,
                    bgColor: c.fg,
                    withTopDivider: !isFirst
                ) {

                    VStack {

                        let isActive = state.activePeriodIndex?.toInt() == periodIndex

                        MyListView__ItemView__RadioView(
                            text: state.periods[periodIndex],
                            isActive: isActive
                        ) {
                            withAnimation {
                                vm.setActivePeriodIndex(index: isActive ? nil : periodIndex.toKotlinInt())
                            }
                        }

                        if isActive {
                            // todo use switch
                            if periodIndex == nil {
                            } else if periodIndex == 0 {
                            } else if periodIndex == 1 {
                                Picker(
                                    "",
                                    selection: $pickerNDaysTemp
                                ) {
                                    // Start from 1 to see WheelPicker
                                    ForEach(1..<667, id: \.self) {
                                        Text("\($0)").tag($0)
                                    }
                                }
                                .pickerStyle(WheelPickerStyle())
                                .onChange(of: pickerNDaysTemp) { newValue in
                                    vm.setSelectedNDays(nDays: newValue.toInt32())
                                }
                                .onAppear {
                                    pickerNDaysTemp = state.selectedNDays.toInt()
                                }
                                // The same problem with .DAY_OF_MONTH
                                // https://stackoverflow.com/a/62603630
                                // Can't decrease the height, the pressing area remains outside the limits
                                // "Add .compositingGroup() after .clipped()" "No longer works on iOS 15.1"
                            } else if periodIndex == 2 {
                                WeekDaysFormView(
                                    weekDays: state.selectedWeekDays,
                                    size: 32,
                                    onChange: { newWeekDays in
                                        vm.upWeekDays(newWeekDays: newWeekDays)
                                    }
                                )
                                .padding(.top, 4)
                                .padding(.bottom, 16)
                                .padding(.leading, MyListView.PADDING_INNER_HORIZONTAL - 1)
                            } else if periodIndex == 3 {
                                let dayNumbers: [Int] = Array(1..<(RepeatingDb.companion.MAX_DAY_OF_MONTH.toInt() + 1))
                                VStack(alignment: .leading, spacing: 8) {
                                    ForEach(dayNumbers.chunked(7), id: \.self) { chunk in
                                        HStack(spacing: 8) {
                                            ForEach(chunk, id: \.self) { day in
                                                let isDaySelected = state.selectedDaysOfMonth.contains(day.toKotlinInt())
                                                DayOfMonthItemView(
                                                    text: "\(day)",
                                                    isDaySelected: isDaySelected
                                                ) {
                                                    vm.toggleDayOfMonth(day: day.toInt32())
                                                }
                                            }
                                        }
                                    }
                                    let isDaySelected = state.selectedDaysOfMonth.contains(RepeatingDb.companion.LAST_DAY_OF_MONTH.toInt().toKotlinInt())
                                    DayOfMonthItemView(
                                        text: "Last Day of the Month",
                                        isDaySelected: isDaySelected,
                                        width: .infinity,
                                        hPaddings: 10
                                    ) {
                                        vm.toggleDayOfMonth(day: RepeatingDb.companion.LAST_DAY_OF_MONTH)
                                    }
                                }
                                .frame(minWidth: 0, maxWidth: .infinity, alignment: .leading)
                                .padding(.top, 6)
                                .padding(.bottom, 16)
                                .padding(.leading, MyListView.PADDING_INNER_HORIZONTAL - 1)
                            } else if periodIndex == 4 {

                                VStack(alignment: .leading, spacing: 0) {

                                    ForEach(state.selectedDaysOfYear, id: \.hash) { monthDayData in

                                        HStack {

                                            Button(
                                                action: {
                                                    vm.delDayOfTheYear(item: monthDayData)
                                                },
                                                label: {
                                                    Image(systemName: "minus.circle.fill")
                                                        .foregroundColor(.red)
                                                }
                                            )

                                            Text(monthDayData.getTitle(isShortOrLong: false))
                                        }
                                        .padding(.bottom, 10)
                                    }

                                    Button(
                                        action: {
                                            isAddDayOfYearSheetPresented = true
                                        },
                                        label: {
                                            Text("Add")
                                        }
                                    )
                                }
                                .frame(minWidth: 0, maxWidth: .infinity, alignment: .leading)
                                .padding(.leading, MyListView.PADDING_INNER_HORIZONTAL)
                                .padding(.bottom, 14)
                                .sheetEnv(isPresented: $isAddDayOfYearSheetPresented) {
                                    AddDayOfYearSheet(
                                        isPresented: $isAddDayOfYearSheetPresented
                                    ) { monthDayItem in
                                        vm.addDayOfTheYear(item: monthDayItem)
                                    }
                                }
                            } else {
                                fatalError()
                            }
                        }
                    }
                }
            }

            Spacer()
        }
        .background(c.bg)
    }
}

private struct DayOfMonthItemView: View {

    let text: String
    let isDaySelected: Bool
    var width: CGFloat = 28
    var height: CGFloat = 28
    var hPaddings: CGFloat = 0
    let onClick: () -> Void

    var body: some View {

        Button(
            action: {
                onClick()
            },
            label: {
                Text("\(text)")
                    .foregroundColor(isDaySelected ? .white : .primary)
                    .frame(width: width, height: height)
                    .font(.system(size: 14))
                    .padding(.horizontal, hPaddings)
            }
        )
        .background(
            ZStack {
                if (isDaySelected) {
                    roundedShape.fill(.blue)
                } else {
                    roundedShape.stroke(.primary, lineWidth: 1)
                }
            }
        )
    }
}

private struct AddDayOfYearSheet: View {

    @Binding var isPresented: Bool
    let onSelect: (RepeatingDbPeriodDaysOfYear.MonthDayItem) -> Void

    @State private var selectedMonthId = 1.toInt32()
    @State private var selectedDayId = 1.toInt32()

    @State private var pickerDayIds: [Int32] = Array(1...5)

    var body: some View {

        VStack {

            SheetHeaderView(
                onCancel: { isPresented.toggle() },
                title: "Day of the Year",
                doneText: "Add",
                isDoneEnabled: true,
                scrollToHeader: 0
            ) {
                isPresented = false
                onSelect(
                    RepeatingDbPeriodDaysOfYear.MonthDayItem(
                        monthId: selectedMonthId,
                        dayId: selectedDayId
                    )
                )
            }
            .padding(.bottom, 20)

            MyListView__Padding__SectionHeader()

            MyListView__ItemView(
                isFirst: true,
                isLast: true
            ) {

                Picker(
                    "",
                    selection: $selectedMonthId
                ) {
                    ForEach(RepeatingDbPeriodDaysOfYear.companion.months, id: \.id) { month in
                        Text(month.getName())
                            .tag(month.id)
                    }
                }
                .pickerStyle(WheelPickerStyle())
                .onChange(of: selectedMonthId) { _ in
                    withAnimation {
                        selectedDayId = 1
                    }
                    upPickerDays()
                }
            }

            MyListView__ItemView(
                isFirst: true,
                isLast: true
            ) {

                Picker(
                    "",
                    selection: $selectedDayId
                ) {
                    ForEach(pickerDayIds, id: \.self) {
                        Text("\($0)")
                            .tag($0)
                    }
                }
                .pickerStyle(WheelPickerStyle())
            }

            Spacer()
        }
        .background(c.sheetBg)
        .onAppear {
            upPickerDays()
        }
    }

    private func upPickerDays() {
        let month = RepeatingDbPeriodDaysOfYear.companion.months.first {
            $0.id == selectedMonthId
        }!
        pickerDayIds = Array(month.days.first...month.days.last)
    }
}

File: android_app/src/main/java/me/timeto/app/ui/RepeatingFormPeriodFs.kt
package me.timeto.app.ui

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.animateColorAsState
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.material.Icon
import androidx.compose.material.Text
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import me.timeto.app.*
import me.timeto.app.R
import me.timeto.shared.db.RepeatingDb
import me.timeto.shared.vm.RepeatingFormPeriodVm

@Composable
fun RepeatingFormPeriodFs(
    layer: WrapperView.Layer,
    defaultPeriod: RepeatingDb.Period?,
    onPick: (RepeatingDb.Period?) -> Unit,
) {

    val (vm, state) = rememberVm(defaultPeriod) {
        RepeatingFormPeriodVm(defaultPeriod = defaultPeriod)
    }

    VStack(
        modifier = Modifier
            .fillMaxSize()
            .background(c.bg),
    ) {

        Fs__HeaderAction(
            title = state.title,
            actionText = state.doneText,
            scrollState = null,
            onCancel = {
                layer.close()
            },
            onDone = {
                vm.buildSelectedPeriod { period ->
                    onPick(period)
                    layer.close()
                }
            },
        )

        MyListView__PaddingFirst()

        val periods = state.periods
        periods.forEachIndexed { index, periodTitle ->
            val isFirst = periods.first() == periodTitle
            MyListView__ItemView(
                isFirst = isFirst,
                isLast = periods.last() == periodTitle,
                withTopDivider = !isFirst,
                bgColor = c.fg,
            ) {
                val isActive = index == state.activePeriodIndex

                Column {

                    MyListView__ItemView__RadioView(
                        text = periodTitle,
                        isActive = isActive,
                        bgColor = c.fg,
                    ) {
                        vm.setActivePeriodIndex(if (isActive) null else index)
                    }

                    AnimatedVisibility(
                        visible = isActive
                    ) {
                        when (index) {
                            0 -> {}
                            1 -> {
                                Column(
                                    modifier = Modifier
                                        .padding(
                                            start = H_PADDING,
                                            top = 4.dp,
                                            bottom = 12.dp,
                                        )
                                ) {

                                    // Remember that indexes from zero, UI from 1.
                                    MyPicker(
                                        items = (1..666).toList(),
                                        containerWidth = 80.dp,
                                        containerHeight = 60.dp,
                                        itemHeight = 20.dp,
                                        selectedIndex = state.selectedNDays - 1,
                                        onChange = { index, _ ->
                                            vm.setSelectedNDays(index + 1)
                                        },
                                    )
                                }
                            }
                            2 -> {
                                WeekDaysFormView(
                                    weekDays = state.selectedWeekDays,
                                    size = 30.dp,
                                    modifier = Modifier
                                        .padding(
                                            start = H_PADDING - 2.dp,
                                            top = 4.dp,
                                            bottom = 12.dp,
                                        ),
                                    onChange = { newWeekDays ->
                                        vm.upWeekDays(newWeekDays)
                                    },
                                )
                            }
                            3 -> {
                                Column(
                                    modifier = Modifier
                                        .padding(
                                            start = H_PADDING - 2.dp,
                                            top = 4.dp,
                                            bottom = 6.dp,
                                        )
                                ) {

                                    (1..RepeatingDb.MAX_DAY_OF_MONTH).chunked(7).forEach { days ->
                                        Row {
                                            days.forEach { day ->
                                                val isDaySelected = day in state.selectedDaysOfMonth
                                                DaysOfMonthItemView(
                                                    dayName = day.toString(),
                                                    isSelected = isDaySelected,
                                                ) {
                                                    vm.toggleDayOfMonth(day)
                                                }
                                            }
                                        }
                                    }

                                    val isLastDaySelected =
                                        RepeatingDb.LAST_DAY_OF_MONTH in state.selectedDaysOfMonth
                                    DaysOfMonthItemView(
                                        dayName = "Last Day of the Month",
                                        isSelected = isLastDaySelected,
                                        width = Dp.Unspecified,
                                        paddingValues = PaddingValues(start = 8.dp, end = 8.dp, bottom = 1.dp)
                                    ) {
                                        vm.toggleDayOfMonth(RepeatingDb.LAST_DAY_OF_MONTH)
                                    }

                                }
                            }
                            4 -> {
                                Column(
                                    modifier = Modifier
                                        .padding(
                                            start = H_PADDING - 1.dp,
                                            bottom = 12.dp,
                                        )
                                ) {

                                    /**
                                     * Impossible to use LazyColumn because of nested scroll. todo animation
                                     */
                                    state.selectedDaysOfYear.forEach { item ->

                                        Row(
                                            modifier = Modifier
                                                .padding(bottom = 8.dp),
                                            verticalAlignment = Alignment.Bottom,
                                        ) {

                                            Text(
                                                "•  " + item.getTitle(isShortOrLong = false),
                                                modifier = Modifier
                                                    .padding(start = 1.dp)
                                                    .offset(y = (-1).dp),
                                                color = c.text,
                                                fontSize = 14.sp,
                                            )

                                            Icon(
                                                painterResource(id = R.drawable.sf_xmark_large_light),
                                                contentDescription = "Close",
                                                modifier = Modifier
                                                    .padding(start = 8.dp)
                                                    .offset(y = 1.dp)
                                                    .size(19.dp, 19.dp)
                                                    .clip(roundedShape)
                                                    .clickable {
                                                        vm.delDayOfTheYear(item)
                                                    }
                                                    .padding(4.dp),
                                                tint = c.red
                                            )
                                        }
                                    }

                                    DaysOfTheYearForm { item ->
                                        vm.addDayOfTheYear(item)
                                    }
                                }
                            }
                            else -> throw Exception()
                        }
                    }
                }
            }
        }
    }
}

@Composable
private fun DaysOfTheYearForm(
    onAdd: (RepeatingDb.Period.DaysOfYear.MonthDayItem) -> Unit,
) {
    val months = RepeatingDb.Period.DaysOfYear.months

    val NO_SELECTED_DAY_STRING = "--"

    var selectedMonthIndex by remember { mutableStateOf(0) }
    var selectedDayIndex by remember(selectedMonthIndex) { mutableStateOf(0) }

    /**
     * Zero element - not selected
     */
    val daysListItems: List<String> = remember(selectedMonthIndex) {
        mutableListOf(
            NO_SELECTED_DAY_STRING,
            *months[selectedMonthIndex].days.map { "$it" }.toTypedArray()
        )
    }

    Row(
        modifier = Modifier.padding(top = 6.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {

        MyPicker(
            items = months,
            containerWidth = 110.dp,
            containerHeight = 30.dp,
            itemHeight = 30.dp,
            selectedIndex = selectedMonthIndex,
            onChange = { index, _ ->
                selectedMonthIndex = index
            },
        )

        Box(Modifier.width(8.dp))

        MyPicker(
            items = daysListItems,
            containerWidth = 40.dp,
            containerHeight = 30.dp,
            itemHeight = 30.dp,
            selectedIndex = selectedDayIndex,
            onChange = { index, _ ->
                selectedDayIndex = index
            },
        )

        // todo rememberDir..?
        val isAddEnabled = selectedDayIndex > 0
        val bgAddColor by animateColorAsState(if (isAddEnabled) c.blue else c.textSecondary.copy(0.1f))
        val textAddColor by animateColorAsState(if (isAddEnabled) c.white else c.textSecondary)

        Text(
            text = "ADD",
            modifier = Modifier
                .padding(start = 10.dp)
                .clip(roundedShape)
                .background(bgAddColor)
                .clickable(
                    enabled = isAddEnabled
                ) {
                    onAdd(
                        RepeatingDb.Period.DaysOfYear.MonthDayItem(
                            monthId = months[selectedMonthIndex].id,
                            dayId = daysListItems[selectedDayIndex].toInt()
                        )
                    )
                    selectedMonthIndex = 0
                    // The day will not "reset" recursively if the month is already 0
                    selectedDayIndex = 0
                }
                .height(30.dp)
                .padding(horizontal = 10.dp)
                .wrapContentHeight(), // To center vertical
            color = textAddColor,
            fontSize = 14.sp,
            fontWeight = FontWeight.W600,
        )
    }
}

@Composable
private fun DaysOfMonthItemView(
    dayName: String,
    isSelected: Boolean,
    width: Dp = 26.dp, // It can be Dp.Unspecified
    height: Dp = 26.dp, // It can be Dp.Unspecified
    paddingValues: PaddingValues = PaddingValues(),
    onClick: () -> Unit,
) {
    val bgColor = animateColorAsState(if (isSelected) c.blue else c.sheetBg)
    Text(
        dayName,
        modifier = Modifier
            .padding(end = 8.dp, bottom = 8.dp)
            .size(width = width, height = height)
            .border(
                width = 1.dp,
                color = if (isSelected) c.blue else c.text,
                shape = roundedShape
            )
            .clip(roundedShape)
            .background(bgColor.value)
            .clickable {
                onClick()
            }
            .padding(paddingValues)
            .wrapContentHeight(), // To center vertical
        color = if (isSelected) c.white else c.text,
        textAlign = TextAlign.Center,
        fontSize = 13.sp,
    )
}
